### PR TITLE
[c++] Add CPU version of standard R-squared metric

### PR DESCRIPTION
**Pull request by nicklamiller**
Contributes to: #6983

- Adds R-squared to regression metrics as `R2Metric`
   - Since this metric measures goodness of fit i.e. higher is better, I made it a subclass of [`Metric`](https://github.com/microsoft/LightGBM/blob/8d7b6f9286d0e6f8f0b6075e08da030e65f8a2cc/include/LightGBM/metric.h#L24-L63) instead of a subclass of [`RegressionMetric`](https://github.com/microsoft/LightGBM/blob/8d7b6f9286d0e6f8f0b6075e08da030e65f8a2cc/src/metric/regression_metric.hpp#L22-L116) (like the other regression metrics, which measure loss/error) and used a very similar approach to how `AveragePrecisionMetric` is defined in `binary_metric.hpp`:
https://github.com/microsoft/LightGBM/blob/8d7b6f9286d0e6f8f0b6075e08da030e65f8a2cc/src/metric/binary_metric.hpp#L270-L385
- For this PR I decided to skip the CUDA implementation, but planned to follow-up with that after this gets merged as part of #6983

### CODE REVIEW

File: src/metric/regression_metric.hpp [hunk 1]
```diff
@@ -318,5 +318,115 @@ class TweedieMetric : public RegressionMetric<TweedieMetric> {
...
+      sum_weights_ = static_cast<double>(num_data_);
+      #pragma omp parallel for num_threads(OMP_NUM_THREADS()) schedule(static) reduction(+:sum_label)
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        sum_label += label_[i];
+      }
+    } else {
+      double local_sum_weights = 0.0f;
+      #pragma omp parallel for num_threads(OMP_NUM_THREADS()) schedule(static) reduction(+:local_sum_weights, sum_label)
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        local_sum_weights += weights_[i];
```
Comment (nicklamiller) [inline src/metric/regression_metric.hpp:349]: Just wanted to give a heads up, I originally updated the `sum_weights_` data member directly in this for-loop, but this resulted in CI failures on some of the R-package jobs for several builds on Windows with the following error ([workflow run](https://github.com/microsoft/LightGBM/actions/runs/17145434917), [line with specific error](https://github.com/microsoft/LightGBM/actions/runs/17145434917/job/48640819518#step:9:571)):

```sh
... error C3028: 'LightGBM::R2Metric::sum_weights_': only a variable or static data member can be used in a data-sharing clause ...
```
So now, the `local_sum_weights` variable is used in the pragma reduction clause and is updated in the loop, and then assigned to `sum_weights_` below. I also had to do this for the `total_sum_squares_` member by introducing the `local_total_sum_squares` variable.

Comment (jameslamb) [reply src/metric/regression_metric.hpp:349]: Thanks for the explanation, makes sense to me! I can't think of a better way to do this, I think this small allocation is totally fine.

**Review by jameslamb — APPROVED** [review]
blegh I thought I submitted a review yesterday but guess I forgot to click "submit review"!

Thank you so much for this excellent PR! It's rare that I review a PR this large and have 0 comments... but I have 0 comments. You addressed everything I would have asked for... followed the project's style, added lightweight but also very effective tests, updated the docs and that list in the R package.

This is just an awesome PR, and I'm excited to add this to LightGBM. Thanks for your hard work!

**Comment by jameslamb** [comment]
I'm not that familiar with C++, so let's see if we can get one other reviewer to look... @borchero could you help us with a review here?

File: src/metric/metric.cpp [hunk 2]
```diff
@@ -76,6 +76,9 @@ Metric* Metric::CreateMetric(const std::string& type, const Config& config) {
       return new CUDAGammaDevianceMetric(config);
     } else if (type == std::string("tweedie")) {
       return new CUDATweedieMetric(config);
+    } else if (type == std::string("r2")) {
+      Log::Warning("Metric R-squared is not implemented in cuda version. Fall back to evaluation on CPU.");
```
Comment (jameslamb) [inline src/metric/metric.cpp]:
```suggestion
      Log::Warning("Metric r2 is not implemented in cuda version. Fall back to evaluation on CPU.");
```

All of the other warning messages in this block match the exact spelling users would pass through `params`. We should follow that with this warning as well.

**Review by jameslamb — APPROVED** [review]
I'd been hoping for another reviewer but looks like no one is available, and I'd been waiting to see valgrind results but I realize we don't need those because this PR isn't adding R-package tests anyway.

I'm confident enough in my ability to review these changes that I think we should just merge this.

On a re-review I left one more small suggestion. I'm going to just apply that and then merge this if/when CI passes.

Thanks again for the excellent contribution!

File: src/metric/metric.cpp (hunk 2 above)
Comment (nicklamiller) [reply src/metric/metric.cpp]: Ah, good catch, thanks for fixing that.

**Review by jameslamb — APPROVED** [review]

**Comment by jameslamb** [comment]
I think the "Optional checks" workflow is going to fail ([like this](https://github.com/microsoft/LightGBM/actions/runs/17786444000/job/50866188989?pr=7008)) until we get a successful run of the valgrind workflow on this branch.... because I put up https://github.com/microsoft/LightGBM/pull/7008#issuecomment-3231766186 but the job wasn't triggered, for the reasons described in #7012 

Sorry @nicklamiller , hopefully we'll be able to get #7035 or something similar merged soon and then re-run that workflow.

File: tests/python_package_test/test_engine.py [hunk 3]
```diff
@@ -4049,6 +4056,27 @@ def test_average_precision_metric():
...
+    train_data = lgb.Dataset(X, label=y)
+    est = lgb.train(
+        params, train_data, num_boost_round=1, valid_sets=[train_data], callbacks=[lgb.record_evaluation(res)]
+    )
+    r2 = res["training"]["r2"][-1]
+    pred = est.predict(X)
+    sklearn_r2 = r2_score(y, pred)
+    assert r2 == pytest.approx(sklearn_r2)
+    # test that R2 is 1 when y has no variance and the model predicts perfectly
+    y = y.copy()
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py:4075]: Why is this line needed?

File: tests/python_package_test/test_engine.py [hunk 4]
```diff
@@ -4049,6 +4056,27 @@ def test_average_precision_metric():
...
+    params = {"objective": "regression", "metric": "r2", "verbose": -1}
+    res = {}
+    train_data = lgb.Dataset(X, label=y)
+    est = lgb.train(
+        params, train_data, num_boost_round=1, valid_sets=[train_data], callbacks=[lgb.record_evaluation(res)]
+    )
+    r2 = res["training"]["r2"][-1]
+    pred = est.predict(X)
+    sklearn_r2 = r2_score(y, pred)
+    assert r2 == pytest.approx(sklearn_r2)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py:4071]: Can we also assert here that `r2 != 0` and `r2 != 1` (edge cases in metric implementation)?

File: tests/python_package_test/test_engine.py (hunk 3 above)
Comment (nicklamiller) [reply tests/python_package_test/test_engine.py:4075]: It isn't really, I've removed this in the latest commit.

**Comment by StrikerRUS** [comment]
Hmm... Errors in CI jobs looks very strange and unrelated to the PR. However, we don't have these errors in `master` branch.

@nicklamiller Could you please take a look?

File: tests/python_package_test/test_engine.py (hunk 3 above)
Comment (nicklamiller) [reply tests/python_package_test/test_engine.py:4075]: @StrikerRUS I stand corrected, removing the `y.copy()` here was the culprit of the odd CI failures [mentioned below](https://github.com/microsoft/LightGBM/pull/7008#issuecomment-3391320128). If we don’t specify `y.copy()`, we reference a view of `y` when we slice via `y[:] = 1`, and according to NumPy modifying a view will update the original, underlying data ([source](https://numpy.org/doc/stable/user/absolute_beginners.html#array-fundamentals)):

> One major difference is that slice indexing of a list copies the elements into a new list, but slicing an array returns a view: an object that refers to the data in the original array. The original array can be mutated using the view.

Sure enough, the tests that were failing were using the same dataset that's used in this test (from `make_synthetic_regression()`)  and were therefore working with modified data.

I successfully ran (most, some still running) CI jobs in [my fork](https://github.com/nicklamiller/LightGBM/pull/2) with these changes as a test, so hopefully we're good to merge once CI finishes here.

**Comment by StrikerRUS** [comment]
@nicklamiller So it looks like `y = y.copy()` is the cause of strange unrelated CI errors, right?..

**Comment by nicklamiller** [comment]
> So it looks like y = y.copy() is the cause of strange unrelated CI errors, right?..

@StrikerRUS Removing `y = y.copy()` was the issue, I should've posted my response in this main thread, but please see [my comment above](https://github.com/microsoft/LightGBM/pull/7008#discussion_r2421745887) for more details.

I've added back `y = y.copy()` in [e5b5868](https://github.com/microsoft/LightGBM/pull/7008/commits/e5b58687ad5026102b6b14ada2265d5bcb0cd2f4) and the seemingly unrelated tests that were failing now pass. The [most recent CI run still fails](https://github.com/microsoft/LightGBM/actions/runs/18453831874/job/52571560893#step:3:41) with a new failure, though that appears to be an unrelated, flaky (API rate related) failure:

```sh
 urllib.error.HTTPError: HTTP Error 403: rate limit exceeded
The last reported status from workflow "R valgrind tests" is failure. Commit fixes and rerun the workflow.
```

Since the most recent failure appears to be flaky, I've merged in the master branch and pushed, so hopefully CI passes this time 🤞

**Comment by jameslamb** [comment]
That `optional-checks` failure isn't "flaky". I tried to trigger an optional workflow at https://github.com/microsoft/LightGBM/pull/7008#issuecomment-3231766186 but then saw it fail with the issues described in #7012.

I've put up #7035 attempting to fix that (and making it easier to fix such things in the future).

**Comment by jameslamb** [comment]
> _I've put up https://github.com/microsoft/LightGBM/pull/7035 attempting to fix that (and making it easier to fix such things in the future)._

@StrikerRUS if the latest changes here look ok to you, I'd be ok with you (temporarily!) changing the branch protection to allow merging this PR while `optional-checks` is failing. That way this PR doesn't need to be blocked by #7035.

**Comment by StrikerRUS** [comment]
@nicklamiller Oh, sorry, I missed in resolved thread your perfect description why unrelated tests were failing! Thanks a lot for the investigation and the fix!

@jameslamb 
> changing the branch protection to allow merging this PR

Sure, totally fine!

**Comment by StrikerRUS** [comment]
Close-reopen to fix `license/cla` status.
